File: 第12次/第一题/test1.py
# -*- coding: utf-8 -*-
from __future__ import division
import numpy as np  
from time import time
#设定数量和循环序列
N=5000
b=np.random.random([N,1])

#构造矩阵与循环数列
A = np.zeros([N,N])
a0 = np.random.random([1,N])
#周期延拓
a1 = np.hstack((a0,a0))

#结构
for k in range(0,N):
    A[k,:]=a1[0,N-k:2*N-k]
#X(m)Y(N+n-m)
#直接解方程

time0=time();
x0 = np.linalg.solve(A,b)

# A(k)*X(k)=Y(K)
#使用N点FFT转到频域
Lambda = np.fft.fftn(A[:,0])
#转换为对角阵，array就不用转行向量了
inv_lambda = np.diag(1/Lambda)
#使用IFFT转换频域
x1 = np.fft.ifftn(np.dot(inv_lambda,np.fft.fftn(b)))
time1=time();

#直接求逆运算
#x2=np.dot(np.linalg.pinv(A),b)
#time2=time();
print("timefft",time1-time0);
#太慢了
#print("timeinv",time2-time1);
print('error',np.linalg.norm(x0-x1));

File: 第12次/第一题/test1.m
clear


N=5000;
b=rand(N,1);             %随机生成向量F

A=zeros(N);
a0=rand(N,1);
for k=0:N-1
   A(:,k+1)=circshift(a0,k);        %随机生成循环矩阵D
end


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
time0=cputime;

x_inv=inv(A)*b;             %直接求逆计算



time1=cputime;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


x_gauss=A\b;             %高斯消去法计算

% [L,U,P]=lu(A);
% x_PLU=U\(L\(P*b));             %PLU分解计算


time2=cputime;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


x_fft=ifft(diag(1./fft(A(:,1)))*fft(b));         %利用快速傅里叶变换计算

time3=cputime;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

error=norm(x_inv-x_fft)+norm(x_gauss-x_fft);         %检验误差
time_inv=time1-time0;
time_gauss=time2-time1;
time_fft=time3-time2;

fprintf('直接求逆计算CPU时间为%f\n', time_inv);
fprintf('高斯消去法计算CPU时间为%f\n', time_gauss);
fprintf('FFT计算CPU时间为%f\n', time_fft);

error
